safenode/src/node/api.rs

```diff
@@ -0,0 +1,87 @@
+// Copyright 2023 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use std::time::Duration;
+
+use futures::future::select_all;
+use xor_name::XorName;
+
+use super::error::Result;
+use crate::{
+    network::{Cmd, Query, Request},
+    storage::chunks::Chunk,
+};
+
+use super::Node;
+
+impl Node {
+    pub async fn store_data(&self, data: Chunk) -> Result<()> {
+        info!("stroing data: {:?}", data.address());
+        self.storage.store(&data).await?;
+        let closest_nodes = self.network.get_closest_nodes(data.address().0).await?;
+        let mut sent_futures = Vec::new();
+        for node in closest_nodes {
+            let req = Request::Cmd(Cmd::StoreChunk(data.clone()));
+            let future = Box::pin(tokio::time::timeout(
+                Duration::from_secs(5),
+                self.network.send_request(req, node),
+            ));
+            sent_futures.push(future);
+        }
+
+        let mut responses = Vec::new();
+        while !sent_futures.is_empty() {
+            match select_all(sent_futures).await {
+                (Ok(res), _, remaining_futures) => {
+                    info!("got response for store_date: {res:?}");
+                    responses.push(res?);
+                    sent_futures = remaining_futures;
+                }
+                (Err(timedout_err), ..) => {
+                    return Err(timedout_err.into());
+                }
+            }
+        }
+
+        //todo: check
+
+        Ok(())
+    }
+
+    pub async fn get_data(&self, xor_name: XorName) -> Result<()> {
+        let closest_nodes = self.network.get_closest_nodes(xor_name).await?;
+
+        let mut sent_futures = Vec::new();
+        for node in closest_nodes {
+            let req = Request::Query(Query::GetChunk(xor_name));
+            let future = Box::pin(tokio::time::timeout(
+                Duration::from_secs(5),
+                self.network.send_request(req, node),
+            ));
+            sent_futures.push(future);
+        }
+
+        let mut responses = Vec::new();
+        while !sent_futures.is_empty() {
+            match select_all(sent_futures).await {
+                (Ok(res), _, remaining_futures) => {
+                    info!("got response for get_data: {res:?}");
```

Comment:
loging the whole chunk could be lengthy.
unless the `Debug` formt of `res` only logs the name of the content?

Reply:
Aye, we got `#[debug(skip)]` over the bytes.